File: orang_tua_app/lib/models/absensi.dart
class AbsensiSummary {
  final int hadir;
  final int sakit;
  final int izin;
  final int alfa;
  final int total;

  const AbsensiSummary({
    required this.hadir,
    required this.sakit,
    required this.izin,
    required this.alfa,
    required this.total,
  });

  double get persentaseHadir => total == 0 ? 0 : (hadir / total * 100);

  factory AbsensiSummary.fromJson(Map<String, dynamic> json) => AbsensiSummary(
        hadir: json['hadir'] ?? 0,
        sakit: json['sakit'] ?? 0,
        izin: json['izin'] ?? 0,
        alfa: json['alfa'] ?? 0,
        total: json['total'] ?? 0,
      );
}

class AbsensiRecord {
  final String id;
  final DateTime tanggal;
  final String? waktuHadir;
  final String status;
  final String? mataPelajaranNama;
  final bool qrCodeScanned;
  final String? catatan;

  const AbsensiRecord({
    required this.id,
    required this.tanggal,
    required this.status,
    this.waktuHadir,
    this.mataPelajaranNama,
    this.qrCodeScanned = false,
    this.catatan,
  });

  factory AbsensiRecord.fromJson(Map<String, dynamic> json) => AbsensiRecord(
        id: json['id'] ?? '',
        tanggal: DateTime.tryParse(json['tanggal'] ?? '') ?? DateTime.now(),
        status: json['status'] ?? 'alfa',
        waktuHadir: json['waktu_hadir'],
        mataPelajaranNama: json['jadwalPelajaran']?['mataPelajaran']?['nama'],
        qrCodeScanned: json['qr_code_scanned'] ?? false,
        catatan: json['catatan'],
      );
}

File: backend/src/models/Absensi.ts
import { DataTypes, Model, Optional } from 'sequelize';
import sequelize from '../config/database';

export type StatusAbsensi = 'hadir' | 'sakit' | 'izin' | 'alfa';

interface AbsensiAttributes {
  id: string;
  siswa_id: string;
  jadwal_pelajaran_id: string;
  tanggal: Date;
  waktu_hadir: Date | null;
  status: StatusAbsensi;
  qr_code_scanned: boolean;
  input_code: string | null;
  catatan: string | null;
  latitude: number | null;
  longitude: number | null;
  created_by: string;
  created_at?: Date;
}

interface AbsensiCreationAttributes extends Optional<AbsensiAttributes, 'id' | 'waktu_hadir' | 'qr_code_scanned' | 'input_code' | 'catatan' | 'latitude' | 'longitude'> {}

class Absensi extends Model<AbsensiAttributes, AbsensiCreationAttributes> implements AbsensiAttributes {
  declare id: string;
  declare siswa_id: string;
  declare jadwal_pelajaran_id: string;
  declare tanggal: Date;
  declare waktu_hadir: Date | null;
  declare status: StatusAbsensi;
  declare qr_code_scanned: boolean;
  declare input_code: string | null;
  declare catatan: string | null;
  declare latitude: number | null;
  declare longitude: number | null;
  declare created_by: string;
  declare readonly created_at: Date;
}

Absensi.init(
  {
    id: { type: DataTypes.UUID, defaultValue: DataTypes.UUIDV4, primaryKey: true },
    siswa_id: { type: DataTypes.UUID, allowNull: false, references: { model: 'siswa', key: 'id' } },
    jadwal_pelajaran_id: { type: DataTypes.UUID, allowNull: false, references: { model: 'jadwal_pelajaran', key: 'id' } },
    tanggal: { type: DataTypes.DATEONLY, allowNull: false },
    waktu_hadir: { type: DataTypes.DATE, allowNull: true },
    status: { type: DataTypes.ENUM('hadir', 'sakit', 'izin', 'alfa'), allowNull: false },
    qr_code_scanned: { type: DataTypes.BOOLEAN, defaultValue: false },
    input_code: { type: DataTypes.STRING(6), allowNull: true },
    catatan: { type: DataTypes.TEXT, allowNull: true },
    latitude: { type: DataTypes.DOUBLE, allowNull: true },
    longitude: { type: DataTypes.DOUBLE, allowNull: true },
    created_by: { type: DataTypes.UUID, allowNull: false },
  },
  {
    sequelize,
    tableName: 'absensi',
    timestamps: true,
    updatedAt: false,
    createdAt: 'created_at',
  }
);

export default Absensi;
